[docs] Create newsletter_september2023.md (#732)
chanzuckerberg/cellxgene-census: docs/articles/2023/newsletter_september2023.md

# Introducing Schema V1.1.0: Key Updates in Census API and Data Layer

*Published: 25 August 2023*  

> *By: [Author Name](mailto:author1@chanzuckerberg.com), [Co-Author](mailto:author2@chanzuckerberg.com)*

Author: ```suggestion
*By: [Maximilian Lombardo](mailto:mlombardo@chanzuckerberg.com)*
```

The Census team is thrilled to announce the introduction of Census schema V1.1.0. This version offers enhanced metadata and API functionalities, tailored to empower your single-cell research.

These features are currently experimental and exclusive to the "latest" version of Census. We invite your feedback as you explore these novel functionalities.

## Schema Changes in V1.1.0

> ### Added `dataset_version_id` to Census Table

Reviewer: ```suggestion
### Added `dataset_version_id` to the datasets data frame
```

The new field `dataset_version_id` has been introduced in `census_obj["census_info"]["datasets"]` to facilitate dataset versioning and management.

> ### Added New Library-Size Normalized Layer in X["normalized"]

Reviewer: ```suggestion
### Added new library-size normalized layer
```

> We've introduced a library-size normalized layer in X["normalized"]. A small sigma value has been consistently added to each position, ensuring that no explicit zeros are recorded.

Reviewer: Please include a sentence that reflects the following message.

> The normalized layer is effectively a division for all values in the count matrix by the corresponding row  (cell) sum of counts. As such, numerical precision of a division can always be impacted, and in addition we introduced a reduction in precision to reduce the disk size required to save these values. As a consequence of that and contrary to the expectation, the row sums of the normalized layer is not guaranteed to be 1.0, but rather slightly lower or greater by a few decimal points.
>
> A small sigma value [...]

> We've introduced a library-size normalized layer in X["normalized"]. A small sigma value has been consistently added to each position, ensuring that no explicit zeros are recorded.

Reviewer: ```suggestion
We've introduced a library-size normalized layer for the RNA measurements of both the human and mouse experiments under `X["normalized"]`. [TEXT FROM COMMENT ABOVE]. A small sigma value has been added during the calculation of the normalized layer to ensure that all non-zero values from the count matrix are also non-zero values in the normalized layer.
```

Author: ```suggestion
We've introduced a library-size normalized layer for the RNA measurements of both the human and mouse experiments under `X["normalized"]`. The normalized layer divides each value in the count matrix by its corresponding row sum (i.e. size normalization), leading to potential numerical precision issues and slight deviations from the expected row sum of 1.0. A small sigma value has been added during the calculation of the normalized layer to ensure that all non-zero values from the count matrix are also non-zero values in the normalized layer.
```

> ### Enhanced Metadata Fields in `ms["RNA"].var` DataFrame

Reviewer: ```suggestion
### Enhanced gene metadata
```

> The `ms["RNA"].var` DataFrame has been enriched with two new metadata fields: `nnz` and `n_measured_obs`, which provide a count of non-zero values and "measured" cells, respectively.

Reviewer: ```suggestion
The `ms["RNA"].var` data frame for both the human and mouse experiments, has been enriched with two new metadata fields: `nnz` and `n_measured_obs`, which provide a count of non-zero values and "measured" cells, respectively.
```

> ### Enhanced Metadata Fields in `ms["RNA"].obs` DataFrame

Reviewer: ```suggestion
### Enhanced cell metadata
```

> The `obs` DataFrame is now augmented with the following new metadata, allowing users to forego common calculations used in early data pre-processing:

Reviewer: ```suggestion
The `obs` data frame for both the human and mouse experiments is now augmented with the following new metadata, allowing users to forego common calculations used in early data pre-processing. For each cell:
```

- `raw_sum`: Represents the count sum derived from X["raw"]
- `nnz`: Enumerates the number of non-zero (nnz) values
- `raw_mean`: Provides the average count of nnz values
- `raw_variance`: Indicates the count variance of nnz values
- `n_measured_vars`: Enumerates the "measured" genes, determined by the sum of the presence matrix

> ## How to Use the New Features

Reviewer: ```suggestion
## How to use the new features
```

> ### Exporting Normalized Data

Reviewer: ```suggestion
### Exporting normalized data
```

> Normalized data can be exported into AnnData with the following code:

Reviewer: ```suggestion
Normalized data can be exported into AnnData specifying the `X_name = "normalized"` argument of the `cellxgene.get_anndata()` method:
```

```python
import cellxgene_census

> with cellxgene_census.open_soma() as census:

Reviewer: This code is not going to work because the latest version of the Census data is not being opened.
```suggestion
with cellxgene_census.open_soma(census_version = "latest") as census:
```

    adata = cellxgene_census.get_anndata(
        census = census,
        organism = "Homo sapiens",
        var_value_filter = "feature_id in ['ENSG00000161798', 'ENSG00000188229']",
        obs_value_filter = "sex == 'female' and cell_type in ['microglial cell', 'neuron']",
        column_names = {"obs": ["assay", "cell_type", "tissue", "tissue_general", "suspension_type", "disease"]},
        X_name = "normalized" # Specificy the normalized layer for this query
        
    )
    
    print(adata)

#Adata with normalized layer
```

> 

Reviewer: Please add SingleCellExperiment and Seurat examples

### Accessing Library-Size Normalized Data Layer via TileDB-SOMA 

> For memory-efficient data retrieval, you can use TileDB-SOMA as outlined below:

Reviewer: Please explain a bit more what's the workflow is here:

1. Perform an axis query with the specified filter.
2. Obtain an iterator of the X matrix and specifically the normalized layer

       measurement_name = "RNA",
       obs_query = tiledbsoma.AxisQuery(
           value_filter = "tissue == 'brain' and sex == 'male'"
       )...
    )

    # Set iterable
    iterator = query.X("normalized").tables()
    
    # Iterate over the matrix count. Get an iterative slice as pyarrow.Table
    raw_slice = next (iterator)

    # close the query
    query.close()

```

### Utilizing Pre-Calculated Stats for Querying `obs` and `var`

To filter based on pre-calculated statistics and export to AnnData, execute the following:

```python
# Example code to fetch all cells with more than 5 detected genes as an AnnData object

import cellxgene_census

with cellxgene_census.open_soma() as census:
    adata = cellxgene_census.get_anndata(
        census = census,
        organism = "Homo sapiens",
        var_value_filter = "feature_id in ['ENSG00000161798', 'ENSG00000188229']",
        obs_value_filter = "nnz > '5' and sex == 'female' and cell_type in ['microglial cell', 'neuron']", 
        column_names = {"obs": ["assay", "cell_type", "tissue", "tissue_general", "suspension_type", "disease"]},
    )

    print(adata)

#adata with filtered cells
```

> 

Reviewer: Please include equivalent examples for Seurat and SingleCellExperiment

> ---

Reviewer: ```suggestion
## Help us improve these data additions
```

We encourage you to engage with these new features in the Census API and share your feedback. This input is invaluable for the ongoing enhancement of the Census project.

For further information on the new library-size normalized layer, please reach out to us at [cellxgene@chanzuckerberg.com](cellxgene@chanzuckerberg.com). To report issues or for additional feedback, refer to our [Census GitHub repository](https://github.com/chanzuckerberg/cellxgene-census/issues).

> ---

Reviewer: ```suggestion
```
